Ajustes solicitados - Relatório de Votações Nominais

cristian-longhi opened the pull request:
<!--- Forneça um resumo geral das suas alterações no título acima -->

## Descrição
<!--- Descreva suas alterações detalhadamente -->
Ajustes/alterações solicitados no PR #3783
## _Issue_ Relacionada

#3720 
<!--- Este projeto apenas aceita _pull requests_ relacionadas à _issues_ abertas. -->
<!--- Se está sugerindo uma nova _feature_ ou mudança, por favor discuta em uma _issue_ antes. -->
<!--- Se está corrigindo um _bug_, deve haver uma _issue_ descrevendo-o com passos para reproduzir. -->
<!--- Por favor, adicione o link para a _issue_ aqui: -->

## Motivação e Contexto
<!--- Por que essa mudança é necessária? Qual problema ela resolve? -->

## Como Isso Foi Testado?
<!--- Por favor, descreva detalhadamente como você testou suas mudanças. -->
<!--- Inclua detalhes do seu ambiente de teste e os testes que você executou -->
<!--- para ver como a sua alteração afeta outras áreas do código, etc. -->

## Capturas de Tela (se apropriado):
<!--- Insira imagens, se apropriado, da adição e/ou correção que foi feita -->

## Tipos de Mudanças
<!--- Quais os tipos de alterações introduzidos pelo seu código? Coloque um `x` em todas as caixas que se aplicam: -->
- [ ] _Bug fix_ (alteração que corrige uma _issue_ e não altera funcionalidades já existentes)
- [x] Nova _feature_ (alteração que adiciona uma funcionalidade e não altera funcionalidades já existentes)
- [ ] Alteração disruptiva (_Breaking change_) (Correção ou funcionalidade que causa alteração nas funcionalidades existentes)

## Checklist:
<!--- Passe por todos os pontos a seguir e coloque um `x` em todas as caixas que se aplicam. -->
<!--- Se você não tem certeza sobre nenhum destes, não hesite em perguntar. Nós estamos aqui para ajudar! -->
- [ ] Eu li o documento de Contribuição (**CONTRIBUTING**).
- [x] Meu código segue o estilo de código deste projeto.
- [ ] Minha alteração requer uma alteração na documentação.
- [ ] Eu atualizei a documentação de acordo.
- [ ] Eu adicionei testes para cobrir minhas mudanças.
- [ ] Todos os testes novos e existentes passaram.

Commit:
Implementa Relatório de Votações Nominais

Commit:
Alterações solicitadas - Relatório de Votações Nominais

edwardoliveira commented on sapl/templates/materia/materialegislativa_filter.html:
```diff
@@ -215,6 +243,14 @@ <h3>{% trans 'Pesquisa concluída com sucesso! Foi encontrada 1 matéria.'%}</h3
+        event.preventDefault();
+      });
+    });
+  });
```
Eu pensei ter feito esta recomendação no PR anterior, mas me enganei.

O jQuery pode simplificar esse trecho:

```suggestion
  $( document ).ready(function() {
      $('.link_votacao_nominal').on('click', function(event) {
        event.preventDefault();
      });
  });
```

edwardoliveira commented on sapl/utils.py:
```diff
@@ -1707,3 +1707,125 @@ def render_to_xlsx(self, context):
         return response
+
+
+class VotacoesMultiFormatOutputMixin(MultiFormatOutputMixin):
```
Acabei de subir uma refatoração de toda essa parte de botões para gerar csv/json/xlsx: https://github.com/interlegis/sapl/commit/b9fe48bc35683df36d99360a8540da3330ba159a

**Tive que fazer algumas concessões como planificar os dados, pois tratam-se de informações tabulares, mas se tiver faltando alguma coisa, tiver algum dúvida ou algum bug, favor me avisar. Favor proceder com a refatoração deste código de acordo...**

Eu ainda estou explorando formas de melhorar os métodos auxiliares, nem que seja em termos de legibilidade e ergonomia pro desenvolvedor ou adicionar documentação, mas testei em normas, pautas, sessões, documentos administrativos e sessão plenária (use-os como exemplo). 

Também percebi que tem uns furos nas lógicas auxiliares originais, tipo na recuperação de headers, mas preferi ter uma versão funcional primeiro e depois endereçar esses outros problemas. 

Ah, tem outra coisa. Não estou muito certo que esse tipo de Mixin especializado tenha que estar em `utils.py`. Talvez seja melhor deixar em `utils.py` só `MultiFormatOutputMixin` e classes/funções diretas, ou então criar um novo arquivo `reports_utils.py` só pra armazenar todas essas coisas relacionadas. Algo a se pensar...

Como exemplo, a refatoração transformou nisso:


```
class PautaMultiFormatOutputMixin(MultiFormatOutputMixin):

    def __mutate_context(self, context):
        context_materias = {
            'object_list': context.get('materias_ordem', []) + context.get('materia_expediente', [])
        }
        return context_materias

    def render_to_csv(self, context):
        return super().render_to_csv(self.__mutate_context(context))

    def render_to_json(self, context):
        return super().render_to_json(self.__mutate_context(context))

    def render_to_xlsx(self, context):
        return super().render_to_xlsx(self.__mutate_context(context))
```

O que antes era isso aqui:


```
class PautaMultiFormatOutputMixin(MultiFormatOutputMixin):

    def render_to_csv(self, context):
        response = HttpResponse(content_type='text/csv')
        response['Content-Disposition'] = f'attachment; filename="sapl_{self.request.resolver_match.url_name}.csv"'
        response['Cache-Control'] = 'no-cache'
        response['Pragma'] = 'no-cache'
        response['Expires'] = 0
        writer = csv.writer(response, delimiter=";",
                            quoting=csv.QUOTE_NONNUMERIC)

        writer.writerow(['Pauta da ' + str(context['sessaoplenaria'])])
        writer.writerow('')

        for item in self.item_context:
            if item[0] in context:

                index = self.item_context.index(item)
                writer.writerow([self.item_context[index][1]])

                data = [[list(self._headers(self.fields_report['csv'][index]))], ]
                for obj in context.get(item[0]):
                    wr = list(self._write_row(obj, self.fields_report['csv'][index]))
                    if wr[0] != data[-1][0][0]:
                        data.append([wr])
                    else:
                        data[-1].append(wr)

                for mri, multirows in enumerate(data):
                    if len(multirows) == 1:
                        writer.writerow(multirows[0])
                    else:
                        v = multirows[0]
                        for ri, cols in enumerate(multirows[1:]):
                            for rc, cell in enumerate(cols):
                                if v[rc] != cell:
                                    v[rc] = f'{v[rc]}\r\n{cell}'

                        writer.writerow(v)
                writer.writerow('')

        return response

    def render_to_json(self, context):

        json_metadata = {'sessaoplenaria': str(context['sessaoplenaria'])}
        for item in self.item_context:
            if item[0] in context:
                index = self.item_context.index(item)
                json_metadata.update({item[0]: {}})
                data = []

                for obj in context.get(item[0]):
                    wr = list(self._write_row(obj, self.fields_report['json'][index]))

                    if not data:
                        data.append([wr])
                        continue

                    if wr[0] != data[-1][0][0]:
                        data.append([wr])
                    else:
                        data[-1].append(wr)

                for mri, multirows in enumerate(data):
                    if len(multirows) == 1:
                        try:
                            v = multirows[0]
                        except TypeError:
                            v = str(multirows[0])
                    else:
                        try:
                            v = str(multirows[0])
                        except TypeError:
                            v = multirows[0]
                        for ri, cols in enumerate(multirows[1:]):
                            for rc, cell in enumerate(cols):
                                if v[rc] != cell:
                                    v[rc] = f'{v[rc]}\r\n{cell}'

                    data[mri] = dict(
                        map(lambda i, j: (i[0], j if type(j) in [str, int, list] else str(j)), self.fields_report['json'][index], v))

                json_metadata.update({item[0]:{
                    'headers': dict(
                        map(lambda i, j: (i[0], j), self.fields_report['json'][index], self._headers(self.fields_report['json'][index]))),
                    'results': data}
                })
        response = JsonResponse(json_metadata)
        response['Content-Disposition'] = f'attachment; filename="sapl_{self.request.resolver_match.url_name}.json"'
        response['Cache-Control'] = 'no-cache'
        response['Pragma'] = 'no-cache'
        response['Expires'] = 0

        return response

    def render_to_xlsx(self, context):

        output = io.BytesIO()
        wb = Workbook(output, {'in_memory': True})

        ws = wb.add_worksheet()
        ws.write('A1', 'Pauta da ' + str(context['sessaoplenaria']))
        row = 2

        for item in self.item_context:
            if item[0] in context:
                index = self.item_context.index(item)
                ws.write(row, 0, self.item_context[index][1])
                row += 1
                data = [[list(self._headers(self.fields_report['xlsx'][index]))], ]

                for obj in context.get(item[0]):
                    wr = list(self._write_row(obj, self.fields_report['xlsx'][index]))
                    if wr[0] != data[-1][0][0]:
                        data.append([wr])
                    else:
                        data[-1].append(wr)

                for mri, multirows in enumerate(data):
                    if len(multirows) == 1:
                        for rc, cell in enumerate(multirows[0]):
                            try:
                                ws.write(row, rc, cell)
                            except TypeError:
                                ws.write(row, rc, str(cell))
                        row += 1
                    else:
                        v = multirows[0]
                        for ri, cols in enumerate(multirows[1:]):
                            for rc, cell in enumerate(cols):
                                if v[rc] != cell:
                                    v[rc] = f'{v[rc]}\r\n{cell}'

                        for rc, cell in enumerate(v):
                            ws.write(row, rc, cell)
                        row += 1
                row += 1
        ws.autofit()
        wb.close()

        output.seek(0)

        response = HttpResponse(output.read(
        ), content_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
        response['Content-Disposition'] = f'attachment; filename="sapl_{self.request.resolver_match.url_name}.xlsx"'
        response['Cache-Control'] = 'no-cache'
        response['Pragma'] = 'no-cache'
        response['Expires'] = 0

        output.close()

        return response
```

edwardoliveira commented on sapl/relatorios/views.py:
```diff
@@ -1880,6 +1885,80 @@ def get_context_data(self, **kwargs):
+        'csv': fields_base_report,
+        'xlsx': fields_base_report,
+        'json': fields_base_report,
+    }
```
Com as recentes refatorações, fica assim:

```suggestion
    export_fields = [
        'votacao_id', 'votacao', 'parlamentar__nome_parlamentar', 'voto'
    ]
```

edwardoliveira commented on sapl/relatorios/views.py:
```diff
@@ -1880,6 +1885,80 @@ def get_context_data(self, **kwargs):
+    }
+
+    def get_queryset(self):
+        if 'format' in self.request.GET:
+            self.model = VotoParlamentar
+            order_fields = ['-votacao_id', 'parlamentar']
+        else:
+            self.model = RegistroVotacao
+            order_fields = ['-id']
+        qs = self.model.objects.filter(Q(ordem__tipo_votacao=2)|Q(expediente__tipo_votacao=2)).order_by(*order_fields)
```
Isso parece um `code smell`... Trocar o model da view na chamada do método que recupera o QuerySet?! Se a ideia é variar o QuerySet retornado, evite de mudar o model da classe, isso complica o código e pode introduzir bugs. 

Vejo duas opções (escolha uma):

Opção 1) Tornar `model` uma variável local inicializada como `model = None` ou `model = self.model`. 

Opção 2) colocar o `qs` dentro dos ifs:

```
    query_params = Q(ordem__tipo_votacao=2)|Q(expediente__tipo_votacao=2)
   
    if 'format' in self.request.GET:
        order_fields = ['-votacao_id', 'parlamentar']
        qs = VotoParlamentar.objects.filter(query_params).order_by(*order_fields) 
    else:
        order_fields = ['-id']
        qs = RegistroVotacao.objects.filter(query_params).order_by(*order_fields)
    return qs
```

edwardoliveira replied:
Particularmente, prefiro a segunda opção, mas a primeira é OK.

edwardoliveira commented on sapl/relatorios/forms.py:
```diff
@@ -68,6 +70,73 @@ def __init__(self, *args, **kwargs):
+                val = value.id
+            except AttributeError:
+                val = value
+            return queryset.filter(
+                    eval(f'Q(ordem__materia__{name}={val}) | Q(expediente__materia__{name}={val})')
+                   )
```
Isso me parece arriscado, e um code smell... avaliar um valor externo para construção de um parâmetro de query **em tempo de execução** não parece elegante, nem correto e abre brecha pra ataques de segurança...


```suggestion
            ordem_q = f"ordem__materia__{name}"
            expediente_q = f"expediente__materia__{name}"
            return queryset.filter(Q(**{ordem_q: val})|Q(**{expediente_q: val}))
```

edwardoliveira commented on sapl/relatorios/forms.py:
```diff
@@ -68,6 +70,73 @@ def __init__(self, *args, **kwargs):
+        )
+
+    def ordem_or_expediente(self, queryset, name, value):
+        if value:
+            try:
+                val = value.id
+            except AttributeError:
+                val = value
```
```suggestion
        if value is None:
           return queryset
        value = getattr(value, "pk", value)
```

edwardoliveira committed:
Merge branch '3.1.x' into ajustes_votacoes_nominais

edwardoliveira reviewed (changes requested):
O código ainda precisa de ajustes, mas está bem melhor que a versão anterior. Particularmente, deve-se observar as mudanças pós-refatoração do código de geração de csv/json/excel.

cristian-longhi committed:
Apply suggestions from code review

Todas as sugestões de alteração acatadas.

Co-authored-by: Edward <9326037+edwardoliveira@users.noreply.github.com>

cristian-longhi committed:
Update views.py

Conforme observação sobre o retorno da QuerySet, escolha da opção 2 - colocar o qs dentro dos if's. Também houve a alteração na view, utilizando diretamente a classe genérica MultiFormatOutputMixin. Com a refatoração efetuada, não foi necessário definir uma especificação da mesma.

edwardoliveira reviewed (approved):
LGTM

edwardoliveira commented on sapl/relatorios/forms.py:
```diff
@@ -68,6 +70,71 @@ def __init__(self, *args, **kwargs):
+    def ordem_or_expediente(self, queryset, name, value):
+        if value is None:
+           return queryset
+        value = getattr(value, "pk", value)
+            ordem_q = f"ordem__materia__{name}"
+            expediente_q = f"expediente__materia__{name}"
+            return queryset.filter(Q(**{ordem_q: val})|Q(**{expediente_q: val}))
+        return queryset
```
@cristian-longhi Erro de indentação e typos (`val` na linha 99 ao invés de `value`, dois returns sem if no mesmo nível). Vou subir um HOT-Fix.
